File: qt/printsupport/gen_qprintpreviewwidget.go
package printsupport

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QPrintPreviewWidget__ViewMode int

const (
	QPrintPreviewWidget__SinglePageView  QPrintPreviewWidget__ViewMode = 0
	QPrintPreviewWidget__FacingPagesView QPrintPreviewWidget__ViewMode = 1
	QPrintPreviewWidget__AllPagesView    QPrintPreviewWidget__ViewMode = 2
)

type QPrintPreviewWidget__ZoomMode int

const (
	QPrintPreviewWidget__CustomZoom QPrintPreviewWidget__ZoomMode = 0
	QPrintPreviewWidget__FitToWidth QPrintPreviewWidget__ZoomMode = 1
	QPrintPreviewWidget__FitInView  QPrintPreviewWidget__ZoomMode = 2
)

type QPrintPreviewWidget struct {
	h          uintptr
	isSubclass bool
}

// NewQPrintPreviewWidget constructs a new QPrintPreviewWidget object.
func NewQPrintPreviewWidget(parent *qt.QWidget) *QPrintPreviewWidget {
	g := newQPrintPreviewWidget(QPrintPreviewWidget_new((*QWidget)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

// NewQPrintPreviewWidget2 constructs a new QPrintPreviewWidget object.
func NewQPrintPreviewWidget2(printer *QPrinter) *QPrintPreviewWidget {
	g := newQPrintPreviewWidget(QPrintPreviewWidget_new2(printer.cPointer()))
	g.isSubclass = true
	return g
}

// NewQPrintPreviewWidget3 constructs a new QPrintPreviewWidget object.
func NewQPrintPreviewWidget3() *QPrintPreviewWidget {
	g := newQPrintPreviewWidget(QPrintPreviewWidget_new3())
	g.isSubclass = true
	return g
}

// NewQPrintPreviewWidget4 constructs a new QPrintPreviewWidget object.
func NewQPrintPreviewWidget4(printer *QPrinter, parent *qt.QWidget) *QPrintPreviewWidget {
	g := newQPrintPreviewWidget(QPrintPreviewWidget_new4(printer.cPointer(), (*QWidget)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

// NewQPrintPreviewWidget5 constructs a new QPrintPreviewWidget object.
func NewQPrintPreviewWidget5(printer *QPrinter, parent *qt.QWidget, flags WindowType) *QPrintPreviewWidget {
	g := newQPrintPreviewWidget(QPrintPreviewWidget_new5(printer.cPointer(), (*QWidget)(parent.UnsafePointer()), (int)(flags)))
	g.isSubclass = true
	return g
}

// NewQPrintPreviewWidget6 constructs a new QPrintPreviewWidget object.
func NewQPrintPreviewWidget6(parent *qt.QWidget, flags WindowType) *QPrintPreviewWidget {
	g := newQPrintPreviewWidget(QPrintPreviewWidget_new6((*QWidget)(parent.UnsafePointer()), (int)(flags)))
	g.isSubclass = true
	return g
}

func (this *QPrintPreviewWidget) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QPrintPreviewWidget_MetaObject(this.h)))
}

func (this *QPrintPreviewWidget) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QPrintPreviewWidget_Metacast(this.h, param1_Cstring))
}

func QPrintPreviewWidget_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QPrintPreviewWidget_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QPrintPreviewWidget) ZoomFactor() float64 {
	return (float64)(QPrintPreviewWidget_ZoomFactor(this.h))
}

func (this *QPrintPreviewWidget) Orientation() qt.QPageLayout__Orientation {
	return (qt.QPageLayout__Orientation)(QPrintPreviewWidget_Orientation(this.h))
}

func (this *QPrintPreviewWidget) ViewMode() ViewMode {
	xxxxxxxxx
}

func (this *QPrintPreviewWidget) ZoomMode() ZoomMode {
	xxxxxxxxx
}

func (this *QPrintPreviewWidget) CurrentPage() int {
	return (int)(QPrintPreviewWidget_CurrentPage(this.h))
}

func (this *QPrintPreviewWidget) PageCount() int {
	return (int)(QPrintPreviewWidget_PageCount(this.h))
}

func (this *QPrintPreviewWidget) SetVisible(visible bool) {
	QPrintPreviewWidget_SetVisible(this.h, (bool)(visible))
}

func (this *QPrintPreviewWidget) Print() {
	QPrintPreviewWidget_Print(this.h)
}

func (this *QPrintPreviewWidget) ZoomIn() {
	QPrintPreviewWidget_ZoomIn(this.h)
}

func (this *QPrintPreviewWidget) ZoomOut() {
	QPrintPreviewWidget_ZoomOut(this.h)
}

func (this *QPrintPreviewWidget) SetZoomFactor(zoomFactor float64) {
	QPrintPreviewWidget_SetZoomFactor(this.h, (double)(zoomFactor))
}

func (this *QPrintPreviewWidget) SetOrientation(orientation qt.QPageLayout__Orientation) {
	QPrintPreviewWidget_SetOrientation(this.h, (int)(orientation))
}

func (this *QPrintPreviewWidget) SetViewMode(viewMode ViewMode) {
	QPrintPreviewWidget_SetViewMode(this.h, viewMode)
}

func (this *QPrintPreviewWidget) SetZoomMode(zoomMode ZoomMode) {
	QPrintPreviewWidget_SetZoomMode(this.h, zoomMode)
}

func (this *QPrintPreviewWidget) SetCurrentPage(pageNumber int) {
	QPrintPreviewWidget_SetCurrentPage(this.h, (int)(pageNumber))
}

func (this *QPrintPreviewWidget) FitToWidth() {
	QPrintPreviewWidget_FitToWidth(this.h)
}

func (this *QPrintPreviewWidget) FitInView() {
	QPrintPreviewWidget_FitInView(this.h)
}

func (this *QPrintPreviewWidget) SetLandscapeOrientation() {
	QPrintPreviewWidget_SetLandscapeOrientation(this.h)
}

func (this *QPrintPreviewWidget) SetPortraitOrientation() {
	QPrintPreviewWidget_SetPortraitOrientation(this.h)
}

func (this *QPrintPreviewWidget) SetSinglePageViewMode() {
	QPrintPreviewWidget_SetSinglePageViewMode(this.h)
}

func (this *QPrintPreviewWidget) SetFacingPagesViewMode() {
	QPrintPreviewWidget_SetFacingPagesViewMode(this.h)
}

func (this *QPrintPreviewWidget) SetAllPagesViewMode() {
	QPrintPreviewWidget_SetAllPagesViewMode(this.h)
}

func (this *QPrintPreviewWidget) UpdatePreview() {
	QPrintPreviewWidget_UpdatePreview(this.h)
}

func (this *QPrintPreviewWidget) PaintRequested(printer *QPrinter) {
	QPrintPreviewWidget_PaintRequested(this.h, printer.cPointer())
}

func (this *QPrintPreviewWidget) OnPaintRequested(slot func(printer *QPrinter)) {
	QPrintPreviewWidget_connect_PaintRequested(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QPrintPreviewWidget_PaintRequested
func miqt_exec_callback_QPrintPreviewWidget_PaintRequested(cb intptr_t, printer *QPrinter) {
	gofunc, ok := cgo.Handle(cb).Value().(func(printer *QPrinter))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQPrinter(printer)

	gofunc(slotval1)
}

func (this *QPrintPreviewWidget) PreviewChanged() {
	QPrintPreviewWidget_PreviewChanged(this.h)
}

func (this *QPrintPreviewWidget) OnPreviewChanged(slot func()) {
	QPrintPreviewWidget_connect_PreviewChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QPrintPreviewWidget_PreviewChanged
func miqt_exec_callback_QPrintPreviewWidget_PreviewChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func QPrintPreviewWidget_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QPrintPreviewWidget_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QPrintPreviewWidget_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QPrintPreviewWidget_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QPrintPreviewWidget) ZoomIn1(zoom float64) {
	QPrintPreviewWidget_ZoomIn1(this.h, (double)(zoom))
}

func (this *QPrintPreviewWidget) ZoomOut1(zoom float64) {
	QPrintPreviewWidget_ZoomOut1(this.h, (double)(zoom))
}

func (this *QPrintPreviewWidget) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QPrintPreviewWidget_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QPrintPreviewWidget) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QPrintPreviewWidget_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QPrintPreviewWidget_MetaObject
func miqt_exec_callback_QPrintPreviewWidget_MetaObject(self QPrintPreviewWidget, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QPrintPreviewWidget{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QPrintPreviewWidget) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QPrintPreviewWidget_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QPrintPreviewWidget) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QPrintPreviewWidget_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QPrintPreviewWidget_Metacast
func miqt_exec_callback_QPrintPreviewWidget_Metacast(self QPrintPreviewWidget, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QPrintPreviewWidget{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/printsupport/gen_qprintpreviewwidget.cpp
// +build ignore

#include <QMetaObject>
#include <QObject>
#include <QPaintDevice>
#include <QPrintPreviewWidget>
#include <QPrinter>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWidget>
#include <qprintpreviewwidget.h>
#include "gen_qprintpreviewwidget.h"
class MiqtVirtualQPrintPreviewWidget : public virtual QPrintPreviewWidget {
public:
MiqtVirtualQPrintPreviewWidget(QWidget* parent): QPrintPreviewWidget(parent) {};
MiqtVirtualQPrintPreviewWidget(QPrinter* printer): QPrintPreviewWidget(printer) {};
MiqtVirtualQPrintPreviewWidget(): QPrintPreviewWidget() {};
MiqtVirtualQPrintPreviewWidget(QPrinter* printer, QWidget* parent): QPrintPreviewWidget(printer, parent) {};
MiqtVirtualQPrintPreviewWidget(QPrinter* printer, QWidget* parent, Qt::WindowFlags flags): QPrintPreviewWidget(printer, parent, flags) {};
MiqtVirtualQPrintPreviewWidget(QWidget* parent, Qt::WindowFlags flags): QPrintPreviewWidget(parent, flags) {};

virtual ~MiqtVirtualQPrintPreviewWidget() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QPrintPreviewWidget::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QPrintPreviewWidget_MetaObject(const_cast<MiqtVirtualQPrintPreviewWidget*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QPrintPreviewWidget::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QPrintPreviewWidget::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QPrintPreviewWidget_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QPrintPreviewWidget::qt_metacast(param1);

	}
};
QPrintPreviewWidget* QPrintPreviewWidget_new(QWidget* parent) {
return new MiqtVirtualQPrintPreviewWidget(parent);
}
QPrintPreviewWidget* QPrintPreviewWidget_new2(QPrinter* printer) {
return new MiqtVirtualQPrintPreviewWidget(printer);
}
QPrintPreviewWidget* QPrintPreviewWidget_new3() {
return new MiqtVirtualQPrintPreviewWidget();
}
QPrintPreviewWidget* QPrintPreviewWidget_new4(QPrinter* printer, QWidget* parent) {
return new MiqtVirtualQPrintPreviewWidget(printer, parent);
}
QPrintPreviewWidget* QPrintPreviewWidget_new5(QPrinter* printer, QWidget* parent, int flags) {
return new MiqtVirtualQPrintPreviewWidget(printer, parent, static_cast<Qt::WindowFlags>(flags));
}
QPrintPreviewWidget* QPrintPreviewWidget_new6(QWidget* parent, int flags) {
return new MiqtVirtualQPrintPreviewWidget(parent, static_cast<Qt::WindowFlags>(flags));
}
void QPrintPreviewWidget_virtbase(QPrintPreviewWidget* src
, QWidget** outptr_QWidget
) {
*outptr_QWidget = static_cast<QWidget*>(src);
}
QMetaObject* QPrintPreviewWidget_MetaObject(const QPrintPreviewWidget* self) {
	return (QMetaObject*) self->metaObject();
}
void* QPrintPreviewWidget_Metacast(QPrintPreviewWidget* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QPrintPreviewWidget_Tr(const char* s) {
	QString _ret = QPrintPreviewWidget::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
double QPrintPreviewWidget_ZoomFactor(const QPrintPreviewWidget* self) {
	qreal _ret = self->zoomFactor();
	return static_cast<double>(_ret);
}
int QPrintPreviewWidget_Orientation(const QPrintPreviewWidget* self) {
	QPageLayout::Orientation _ret = self->orientation();
	return static_cast<int>(_ret);
}
ViewMode QPrintPreviewWidget_ViewMode(const QPrintPreviewWidget* self) {
	return self->viewMode();
}
ZoomMode QPrintPreviewWidget_ZoomMode(const QPrintPreviewWidget* self) {
	return self->zoomMode();
}
int QPrintPreviewWidget_CurrentPage(const QPrintPreviewWidget* self) {
	return self->currentPage();
}
int QPrintPreviewWidget_PageCount(const QPrintPreviewWidget* self) {
	return self->pageCount();
}
void QPrintPreviewWidget_SetVisible(QPrintPreviewWidget* self, bool visible) {
	self->setVisible(visible);
}
void QPrintPreviewWidget_Print(QPrintPreviewWidget* self) {
	self->print();
}
void QPrintPreviewWidget_ZoomIn(QPrintPreviewWidget* self) {
	self->zoomIn();
}
void QPrintPreviewWidget_ZoomOut(QPrintPreviewWidget* self) {
	self->zoomOut();
}
void QPrintPreviewWidget_SetZoomFactor(QPrintPreviewWidget* self, double zoomFactor) {
	self->setZoomFactor(static_cast<qreal>(zoomFactor));
}
void QPrintPreviewWidget_SetOrientation(QPrintPreviewWidget* self, int orientation) {
	self->setOrientation(static_cast<QPageLayout::Orientation>(orientation));
}
void QPrintPreviewWidget_SetViewMode(QPrintPreviewWidget* self, ViewMode viewMode) {
	self->setViewMode(viewMode);
}
void QPrintPreviewWidget_SetZoomMode(QPrintPreviewWidget* self, ZoomMode zoomMode) {
	self->setZoomMode(zoomMode);
}
void QPrintPreviewWidget_SetCurrentPage(QPrintPreviewWidget* self, int pageNumber) {
	self->setCurrentPage(static_cast<int>(pageNumber));
}
void QPrintPreviewWidget_FitToWidth(QPrintPreviewWidget* self) {
	self->fitToWidth();
}
void QPrintPreviewWidget_FitInView(QPrintPreviewWidget* self) {
	self->fitInView();
}
void QPrintPreviewWidget_SetLandscapeOrientation(QPrintPreviewWidget* self) {
	self->setLandscapeOrientation();
}
void QPrintPreviewWidget_SetPortraitOrientation(QPrintPreviewWidget* self) {
	self->setPortraitOrientation();
}
void QPrintPreviewWidget_SetSinglePageViewMode(QPrintPreviewWidget* self) {
	self->setSinglePageViewMode();
}
void QPrintPreviewWidget_SetFacingPagesViewMode(QPrintPreviewWidget* self) {
	self->setFacingPagesViewMode();
}
void QPrintPreviewWidget_SetAllPagesViewMode(QPrintPreviewWidget* self) {
	self->setAllPagesViewMode();
}
void QPrintPreviewWidget_UpdatePreview(QPrintPreviewWidget* self) {
	self->updatePreview();
}
void QPrintPreviewWidget_PaintRequested(QPrintPreviewWidget* self, QPrinter* printer) {
	self->paintRequested(printer);
}
void QPrintPreviewWidget_connect_PaintRequested(QPrintPreviewWidget* self, intptr_t slot) {
	MiqtVirtualQPrintPreviewWidget::connect(self, static_cast<void (QPrintPreviewWidget::*)(QPrinter*)>(&QPrintPreviewWidget::paintRequested), self, [=](QPrinter* printer) {
		QPrinter* sigval1 = printer;
		miqt_exec_callback_QPrintPreviewWidget_PaintRequested(slot, sigval1);
	});
}
void QPrintPreviewWidget_PreviewChanged(QPrintPreviewWidget* self) {
	self->previewChanged();
}
void QPrintPreviewWidget_connect_PreviewChanged(QPrintPreviewWidget* self, intptr_t slot) {
	MiqtVirtualQPrintPreviewWidget::connect(self, static_cast<void (QPrintPreviewWidget::*)()>(&QPrintPreviewWidget::previewChanged), self, [=]() {
		miqt_exec_callback_QPrintPreviewWidget_PreviewChanged(slot);
	});
}
struct miqt_string QPrintPreviewWidget_Tr2(const char* s, const char* c) {
	QString _ret = QPrintPreviewWidget::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QPrintPreviewWidget_Tr3(const char* s, const char* c, int n) {
	QString _ret = QPrintPreviewWidget::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QPrintPreviewWidget_ZoomIn1(QPrintPreviewWidget* self, double zoom) {
	self->zoomIn(static_cast<qreal>(zoom));
}
void QPrintPreviewWidget_ZoomOut1(QPrintPreviewWidget* self, double zoom) {
	self->zoomOut(static_cast<qreal>(zoom));
}
void QPrintPreviewWidget_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQPrintPreviewWidget*>( (QPrintPreviewWidget*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QPrintPreviewWidget_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQPrintPreviewWidget*)(self) )->virtualbase_MetaObject();
}
void QPrintPreviewWidget_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQPrintPreviewWidget*>( (QPrintPreviewWidget*)(self) )->handle__Metacast = slot;
}
void* QPrintPreviewWidget_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQPrintPreviewWidget*)(self) )->virtualbase_Metacast(param1);
}
void QPrintPreviewWidget_Delete(QPrintPreviewWidget* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQPrintPreviewWidget*>( self );
	} else {
		delete self;
	}
}
